File: accbot-android/app/src/main/java/com/accbot/dca/domain/usecase/CalculateStrategyMultiplierUseCase.kt
package com.accbot.dca.domain.usecase

import android.util.Log
import com.accbot.dca.data.remote.MarketDataService
import com.accbot.dca.domain.model.DcaStrategy
import com.accbot.dca.domain.model.MarketData
import com.accbot.dca.domain.model.StrategyMultiplierResult
import java.math.BigDecimal
import javax.inject.Inject

/**
 * Use case for calculating DCA purchase multiplier based on strategy
 */
class CalculateStrategyMultiplierUseCase @Inject constructor(
    private val marketDataService: MarketDataService
) {
    companion object {
        private const val TAG = "StrategyMultiplier"
    }

    /**
     * Calculate the purchase amount multiplier based on strategy and current market conditions
     *
     * @param strategy The DCA strategy to use
     * @param crypto The cryptocurrency symbol (BTC, ETH, etc.)
     * @param fiat The fiat currency (EUR, USD, etc.)
     * @return StrategyMultiplierResult containing the multiplier and explanation
     */
    suspend operator fun invoke(
        strategy: DcaStrategy,
        crypto: String,
        fiat: String
    ): StrategyMultiplierResult {
        return when (strategy) {
            is DcaStrategy.Classic -> calculateClassic()
            is DcaStrategy.AthBased -> calculateAthBased(strategy, crypto, fiat)
            is DcaStrategy.FearAndGreed -> calculateFearAndGreed(strategy)
        }
    }

    private fun calculateClassic(): StrategyMultiplierResult {
        return StrategyMultiplierResult(
            multiplier = 1.0f,
            reason = "Classic DCA: Fixed amount purchase"
        )
    }

    private suspend fun calculateAthBased(
        strategy: DcaStrategy.AthBased,
        crypto: String,
        fiat: String
    ): StrategyMultiplierResult {
        val cryptoData = marketDataService.getCryptoData(crypto, fiat)

        if (cryptoData == null) {
            Log.w(TAG, "Could not fetch ATH data for $crypto/$fiat, using default multiplier")
            return StrategyMultiplierResult(
                multiplier = 1.0f,
                reason = "ATH data unavailable, using default amount"
            )
        }

        val athDistance = cryptoData.athDistance

        // Find the appropriate tier
        val multiplier = strategy.tiers
            .sortedBy { it.maxDistancePercent }
            .firstOrNull { athDistance <= it.maxDistancePercent }
            ?.multiplier ?: 1.0f

        val distancePercent = (athDistance * 100).toInt()

        Log.d(TAG, "ATH-based: $crypto is $distancePercent% below ATH, multiplier: $multiplier")

        return StrategyMultiplierResult(
            multiplier = multiplier,
            reason = "$crypto is $distancePercent% below ATH → ${formatMultiplier(multiplier)}",
            marketData = MarketData(
                currentPrice = cryptoData.currentPrice,
                allTimeHigh = cryptoData.allTimeHigh
            )
        )
    }

    private suspend fun calculateFearAndGreed(
        strategy: DcaStrategy.FearAndGreed
    ): StrategyMultiplierResult {
        val fngData = marketDataService.getFearGreedIndex()

        if (fngData == null) {
            Log.w(TAG, "Could not fetch Fear & Greed data, using default multiplier")
            return StrategyMultiplierResult(
                multiplier = 1.0f,
                reason = "Fear & Greed data unavailable, using default amount"
            )
        }

        val index = fngData.value

        // Find the appropriate tier
        val multiplier = strategy.tiers
            .sortedBy { it.maxIndex }
            .firstOrNull { index <= it.maxIndex }
            ?.multiplier ?: 1.0f

        Log.d(TAG, "Fear & Greed: Index is $index (${fngData.classification}), multiplier: $multiplier")

        return StrategyMultiplierResult(
            multiplier = multiplier,
            reason = "${fngData.classification} ($index) → ${formatMultiplier(multiplier)}",
            marketData = MarketData(
                currentPrice = BigDecimal.ZERO,
                fearGreedIndex = index
            )
        )
    }

    private fun formatMultiplier(multiplier: Float): String {
        return when {
            multiplier < 1.0f -> "${(multiplier * 100).toInt()}% of base"
            multiplier == 1.0f -> "normal amount"
            else -> "${(multiplier * 100).toInt()}% of base"
        }
    }
}

File: accbot-ios/AccBot/Domain/UseCase/CalculateStrategyMultiplierUseCase.swift
import Foundation
import os

/// Calculates DCA purchase multiplier based on strategy and current market conditions.
final class CalculateStrategyMultiplierUseCase {
    private let marketDataService: MarketDataService
    private let logger = Logger(subsystem: "com.accbot.dca", category: "StrategyMultiplier")

    init(marketDataService: MarketDataService) {
        self.marketDataService = marketDataService
    }

    func invoke(
        strategy: DcaStrategy,
        crypto: String,
        fiat: String
    ) async -> StrategyMultiplierResult {
        switch strategy {
        case .classic:
            return StrategyMultiplierResult(
                multiplier: 1.0,
                reason: "Classic DCA: Fixed amount purchase"
            )

        case .athBased(let tiers):
            return await calculateAthBased(tiers: tiers, crypto: crypto, fiat: fiat)

        case .fearAndGreed(let tiers):
            return await calculateFearAndGreed(tiers: tiers)
        }
    }

    private func calculateAthBased(tiers: [AthTier], crypto: String, fiat: String) async -> StrategyMultiplierResult {
        guard let currentPrice = await marketDataService.getCurrentPrice(crypto: crypto, fiat: fiat),
              let ath = await marketDataService.getAllTimeHigh(crypto: crypto, fiat: fiat),
              ath > 0
        else {
            logger.warning("Could not fetch ATH data for \(crypto)/\(fiat), using default multiplier")
            return StrategyMultiplierResult(multiplier: 1.0, reason: "ATH data unavailable, using default amount")
        }

        let currentDouble = NSDecimalNumber(decimal: currentPrice).doubleValue
        let athDouble = NSDecimalNumber(decimal: ath).doubleValue
        let athDistance = Float(1.0 - (currentDouble / athDouble))

        let sortedTiers = tiers.sorted { $0.maxDistancePercent < $1.maxDistancePercent }
        let multiplier = sortedTiers.first(where: { athDistance <= $0.maxDistancePercent })?.multiplier ?? 1.0

        let distancePercent = Int(athDistance * 100)
        logger.info("ATH-based: \(crypto) is \(distancePercent)% below ATH, multiplier: \(multiplier)")

        return StrategyMultiplierResult(
            multiplier: multiplier,
            reason: "\(crypto) is \(distancePercent)% below ATH \u{2192} \(formatMultiplier(multiplier))",
            marketData: MarketData(currentPrice: currentPrice, allTimeHigh: ath, fearGreedIndex: nil)
        )
    }

    private func calculateFearAndGreed(tiers: [FearGreedTier]) async -> StrategyMultiplierResult {
        guard let index = await marketDataService.getFearGreedIndex() else {
            logger.warning("Could not fetch Fear & Greed data, using default multiplier")
            return StrategyMultiplierResult(multiplier: 1.0, reason: "Fear & Greed data unavailable, using default amount")
        }

        let sortedTiers = tiers.sorted { $0.maxIndex < $1.maxIndex }
        let multiplier = sortedTiers.first(where: { index <= $0.maxIndex })?.multiplier ?? 1.0

        let classification = FearGreedClassification.label(for: index)
        logger.info("Fear & Greed: Index is \(index) (\(classification)), multiplier: \(multiplier)")

        return StrategyMultiplierResult(
            multiplier: multiplier,
            reason: "\(classification) (\(index)) \u{2192} \(formatMultiplier(multiplier))",
            marketData: MarketData(currentPrice: 0, allTimeHigh: nil, fearGreedIndex: index)
        )
    }

    private func formatMultiplier(_ multiplier: Float) -> String {
        let pct = Int(multiplier * 100)
        if multiplier == 1.0 {
            return "normal amount"
        }
        return "\(pct)% of base"
    }
}
